<h1>importing libraries<h1>

In [1]:
!pip install beautifulsoup4
from bs4 import BeautifulSoup # library for scraping from a website

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium
import folium # map rendering library

!pip install geocoder
import geocoder # to get longitude and latitude

     |████████████████████████████████| 92kB 15.3MB/s eta 0:00:01
     |████████████████████████████████| 102kB 14.5MB/s ta 0:00:01


In [2]:
#assign the link of the website through which we are going to scrape the data and assign it to variable named website_url.
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = requests.get(url).text

In [3]:
#Beautiful Soup is a Python package for parsing HTML and XML documents. 
#It creates a parse tree for parsed pages that can be used to extract data from HTML, which is useful for web scraping
from bs4 import BeautifulSoup
soup = BeautifulSoup(page,'html.parser')
#print(soup.prettify())

In [4]:
# extract the table
My_table = soup.find('table',class_='sortable')
#My_table

In [5]:
#Extracting all the links
links = My_table.findAll('a')
#links

In [6]:
row = [] # initialize row list

for tr in My_table.find_all('tr'):                                          
    if tr.find_all('th') == []:                                           
        row.append([td.get_text(strip=True) for td in tr.find_all('td')])

In [7]:
# Assign columns names and turn row into a dataframe of neighborhoods
column_names = ['Postcode', 'Borough', 'Neighborhood']
# create a raw table of neighborhoods
neighborhoods = pd.DataFrame(row, columns=column_names)
neighborhoods.head(5)

#neighborhoods.shape

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [8]:
# Drop the Not Assigned values for Borough
df = neighborhoods[neighborhoods.Borough!="Not assigned"]
df.reset_index(drop=True, inplace=True) 
df.head(10)
#df.shape

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Not assigned
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern

In [11]:
# using group by function to merge two rows with same Postcode and Borough values in to a single row
df = df.groupby(['Postcode', 'Borough'], as_index=False).agg(lambda x: ', '.join(x))
df.head()

Postcode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [12]:
# If Neighborhood has NotAssigned values then assigning them as same as Borough values
neighborhood_na = df[df['Neighborhood'] == 'Not assigned'].index
df.iloc[neighborhood_na, 2] = df['Borough'][neighborhood_na]

In [13]:
df.head(10)

Postcode      Borough                                     Neighborhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

In [14]:
# shape of the data
df.shape

(103, 3)

<h1>Longitude and Latitude columns <h1>

<h3> For this part I am goind to merge the two DF and making into a new DF. I want to try in two different ways 1) using merge function and 2) using Geocoder package. Geocoder is giving me an error. I will update this github project if I solve the issue.<h3>

In [15]:
# The code was removed by Watson Studio for sharing.

Postcode   Latitude  Longitude
0       M1B  43.806686 -79.194353
1       M1C  43.784535 -79.160497
2       M1E  43.763573 -79.188711
3       M1G  43.770992 -79.216917
4       M1H  43.773136 -79.239476
5       M1J  43.744734 -79.239476
6       M1K  43.727929 -79.262029
7       M1L  43.711112 -79.284577
8       M1M  43.716316 -79.239476
9       M1N  43.692657 -79.264848
10      M1P  43.757410 -79.273304
11      M1R  43.750072 -79.295849
12      M1S  43.794200 -79.262029
13      M1T  43.781638 -79.304302
14      M1V  43.815252 -79.284577
15      M1W  43.799525 -79.318389
16      M1X  43.836125 -79.205636
17      M2H  43.803762 -79.363452
18      M2J  43.778517 -79.346556
19      M2K  43.786947 -79.385975

In [16]:
# merging the two DF's using the merge function
df2 = pd.merge(df, df1, on = 'Postcode')
#df2.head(20)

In [17]:
df2

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.23

<h1> 3. Explore and Cluster the neighborhood in Toronto. <h1>
<h4> I am using complete data set instead of filtering. so, it contains 103 postal codes including repeated ones. <h4>

In [18]:
# I am following lab instructions for this part. # It displays starting latitude and longitude values.
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 43.653963, -79.387207.


In [19]:
# create map of canada using latitude and longitude values
map_canada = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df2['Latitude'], df2['Longitude'], df2['Borough'], df2['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_canada)  
    
map_canada

In [20]:
##Filtering the data set for only Toronto. I took google help to find the filtering way.
# New Dataset with only originaly city. It turns out there are 38 different postalcodes in the original data set.
df3= df2[df2['Borough'].str.find('Toronto') != -1].reset_index(drop=True)
df3.shape

(38, 5)

In [21]:
g = geocoder.arcgis('Toronto, Ontario')
lat_tor = g.latlng[0]
lng_tor = g.latlng[1]

# Create a map of Toronto
map_toronto = folium.Map(location=[lat_tor, lng_tor], zoom_start=11)

# Add markers to map
for lat, lng, bor, postcode in zip(df3['Latitude'], df3['Longitude'], df3['Borough'], df3['Postcode']):
    label = '{}, {}'.format(postcode, bor)        # popup labels with postcode and borough
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker([lat, lng],
                        radius=5,
                        popup=label,
                        color='blue',
                        fill=True,
                        fill_color='#3186cc',
                        fill_opacity=0.7,
                        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h1>Define Foursquare Credentials and Version. <h1>

In [22]:
CLIENT_ID = 'RAJI4VCJFAOL1SD4AWIZBRRHF1UWMLKHI01XNLNOIYQDHCIT' # your Foursquare ID
CLIENT_SECRET = '3XOQZKAPU0JN0XVBQ4NUCJFAS0K2T0B0JLRCY1ALGCIZGHPR' # your Foursquare Secret
VERSION = '20180604'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: RAJI4VCJFAOL1SD4AWIZBRRHF1UWMLKHI01XNLNOIYQDHCIT
CLIENT_SECRET:3XOQZKAPU0JN0XVBQ4NUCJFAS0K2T0B0JLRCY1ALGCIZGHPR


In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):        
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
                CLIENT_ID, 
                CLIENT_SECRET, 
                VERSION, 
                lat, 
                lng, 
                radius, 
                LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(name,
                             lat,
                             lng,
                             v['venue']['name'],
                             v['venue']['location']['lat'],
                             v['venue']['location']['lng'],
                             v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postcode', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [24]:
# Limit the number of venues to 100
LIMIT = 100

# Get the venues for each postcode
Toronto_venues = getNearbyVenues(names=df3['Postcode'],
                                 latitudes=df3['Latitude'],
                                 longitudes=df3['Longitude'],
                                 radius=500) # limit radius to 700 meters

In [25]:
print(Toronto_venues.shape)
Toronto_venues.head(10)

(1699, 7)


Postcode   Latitude  Longitude                               Venue  \
0      M4E  43.676357 -79.293031                   Glen Manor Ravine   
1      M4E  43.676357 -79.293031  The Big Carrot Natural Food Market   
2      M4E  43.676357 -79.293031                 Grover Pub and Grub   
3      M4E  43.676357 -79.293031                       Upper Beaches   
4      M4E  43.676357 -79.293031                          Dip 'n Sip   
5      M4K  43.679557 -79.352188                            Pantheon   
6      M4K  43.679557 -79.352188                        Dolce Gelato   
7      M4K  43.679557 -79.352188                       MenEssentials   
8      M4K  43.679557 -79.352188                     Cafe Fiorentina   
9      M4K  43.679557 -79.352188                          La Diperie   

   Venue Latitude  Venue Longitude      Venue Category  
0       43.676821       -79.293942               Trail  
1       43.678879       -79.297734   Health Food Store  
2       43.679181       -79.297215                 Pub  
3       43.680563       -79.292869        Neighborhood  
4       43.678897       -79.297745         Coffee Shop  
5       43.677621       -79.351434    Greek Restaurant  
6       43.677773       -79.351187      Ice Cream Shop  
7       43.677820       -79.351265      Cosmetics Shop  
8       43.677743       -79.350115  Italian Restaurant  
9       43.677530       -79.352295      Ice Cream Shop

In [26]:
# Let's count how many venue are in each neighborhood
Count_venues = pd.DataFrame()
Count_venues['Postcode'] = Toronto_venues.groupby('Postcode').count().reset_index()['Postcode']
Count_venues['Count'] = Toronto_venues.groupby('Postcode').count().reset_index()['Venue']
Count_venues

Postcode  Count
0       M4E      5
1       M4K     42
2       M4L     18
3       M4M     39
4       M4N      4
5       M4P      9
6       M4R     20
7       M4S     35
8       M4T      4
9       M4V     15
10      M4W      5
11      M4X     45
12      M4Y     84
13      M5A     49
14      M5B    100
15      M5C    100
16      M5E     57
17      M5G     84
18      M5H    100
19      M5J    100
20      M5K    100
21      M5L    100
22      M5N      1
23      M5P      4
24      M5R     24
25      M5S     35
26      M5T    100
27      M5V     14
28      M5W     96
29      M5X    100
30      M6G     15
31      M6H     16
32      M6J     65
33      M6K     21
34      M6P     23
35      M6R     15
36      M6S     36
37      M7Y     19

In [27]:
# Analyze Each Neighborhood
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")


Toronto_onehot['Postcode'] = Toronto_venues['Postcode'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()


Postcode  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0      M4E                  0        0                   0             0   
1      M4E                  0        0                   0             0   
2      M4E                  0        0                   0             0   
3      M4E                  0        0                   0             0   
4      M4E                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0                 0              0          0           0           0       0   
1                 0              0          0           0           0       0   
2                 0              0          0           0           0       0   
3                 0              0          0           0           0       0   
4                 0              0          0           0           0       0   

   Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  Beer Bar  \
0     0    0                 0                   0      0         0   
1     0    0                 0                   0      0         0   
2     0    0                 0                   0      0         0   
3     0    0                 0                   0      0         0   
4     0    0                 0                   0      0         0   

   Beer Store  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0       0              0          0         0   
1           0       0              0          0         0   
2           0       0              0          0         0   
3           0       0              0          0         0   
4           0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0                          0             0                     0            0   
1                          0             0                     0            0   
2                          0             0                     0            0   
3                          0             0                     0            0   
4                          0             0                     0            0   

   Chinese Restaurant  Chocolate 

In [28]:
Toronto_onehot.shape

(1699, 240)

In [29]:
#Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
Toronto_grouped = Toronto_onehot.groupby('Postcode').mean().reset_index()
Toronto_grouped

Postcode  Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0       M4E           0.000000  0.000000            0.000000      0.000000   
1       M4K           0.000000  0.000000            0.000000      0.000000   
2       M4L           0.000000  0.000000            0.000000      0.000000   
3       M4M           0.000000  0.000000            0.000000      0.000000   
4       M4N           0.000000  0.000000            0.000000      0.000000   
5       M4P           0.000000  0.000000            0.000000      0.000000   
6       M4R           0.000000  0.000000            0.000000      0.000000   
7       M4S           0.000000  0.000000            0.000000      0.000000   
8       M4T           0.000000  0.000000            0.000000      0.000000   
9       M4V           0.000000  0.000000            0.000000      0.000000   
10      M4W           0.000000  0.000000            0.000000      0.000000   
11      M4X           0.000000  0.000000            0.000000      0.000000   
12      M4Y           0.011905  0.000000            0.000000      0.000000   
13      M5A           0.000000  0.000000            0.000000      0.000000   
14      M5B           0.000000  0.000000            0.000000      0.000000   
15      M5C           0.000000  0.000000            0.000000      0.000000   
16      M5E           0.000000  0.000000            0.000000      0.000000   
17      M5G           0.000000  0.000000            0.000000      0.000000   
18      M5H           0.000000  0.000000            0.000000      0.000000   
19      M5J           0.000000  0.000000            0.000000      0.000000   
20      M5K           0.000000  0.000000            0.000000      0.000000   
21      M5L           0.000000  0.000000            0.000000      0.000000   
22      M5N           0.000000  0.000000            0.000000      0.000000   
23      M5P           0.000000  0.000000            0.000000      0.000000   
24      M5R           0.000000  0.000000            0.000000      0.000000   
25      M5S           0.000000  0.000000            0.000000      0.000000   
26      M5T           0.000000  0.000000            0.000000      0.000000   
27      M5V           0.000000  0.071429            0.071429      0.071429   
28      M5W           0.000000  0.000000            0.000000      0.000000   
29      M5X           0.000000  0.000000            0.000000      0.000000   
30      M6G           0.000000  0.000000            0.000000      0.000000   
31      M6H           0.000000  0.000000            0.000000      0.000000   
32      M6J           0.000000  0.000000            0.000000      0.000000   
33      M6K           0.000000  0.000000            0.000000      0.000000   
34      M6P           0.000000  0.000000            0.000000      0.000000   
35      M6R           0.000000  0.000000            0.000000      0.000000   
36      M6S           0.000000  0.000000            0.000000      0.000000   
37      M7Y           0.000000  0.000000            0.000000      0.000000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0         0.000000         0.000000          0.000000             0.000000   
1         0.000000         0.000000          0.000000             0.023810   
2         0.000000         0.000000          0.000000             0.000000   
3         0.000000         0.000000          0.000000             0.051282   
4         0.000000         0.000000          0.000000             0.000000   
5         0.000000         0.000000          0.000000             0.000000   
6         0.000000         0.000000          0.000000             0.000000   
7         0.000000         0.000000          0.000000             0.000000   
8         0.000000         0.000000          0.000000             0.000000   
9         0.000000         0.000000          0.000000             0.066667   
10        0.000000         0.000000          0.000000             0.000000   
11        0.000000      

In [30]:
# Confirm the shape
Toronto_grouped.shape

(38, 240)

In [31]:
# sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
Toronto_venues_sorted = pd.DataFrame(columns=columns)
Toronto_venues_sorted['Postcode'] = Toronto_grouped['Postcode']

for ind in np.arange(Toronto_grouped.shape[0]):
    Toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

Toronto_venues_sorted.head()

Postcode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0      M4E          Neighborhood                   Pub     Health Food Store   
1      M4K      Greek Restaurant           Coffee Shop    Italian Restaurant   
2      M4L           Pizza Place    Italian Restaurant             Pet Store   
3      M4M                  Café           Coffee Shop    Italian Restaurant   
4      M4N                  Park              Bus Line    Dim Sum Restaurant   

  4th Most Common Venue   5th Most Common Venue 6th Most Common Venue  \
0           Coffee Shop                   Trail      Doner Restaurant   
1        Ice Cream Shop  Furniture / Home Store           Yoga Studio   
2                   Pub           Movie Theater        Sandwich Place   
3                Bakery     American Restaurant           Yoga Studio   
4           Swim School             Yoga Studio            Donut Shop   

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
0            Donut Shop   Dumpling Restaurant  Eastern European Restaurant   
1             Bookstore               Brewery              Bubble Tea Shop   
2         Burrito Place          Burger Joint         Fast Food Restaurant   
3       Coworking Space    Seafood Restaurant               Sandwich Place   
4     Fish & Chips Shop   Filipino Restaurant         Fast Food Restaurant   

  10th Most Common Venue  
0      Electronics Store  
1           Burger Joint  
2                Brewery  
3            Cheese Shop  
4         Farmers Market

In [33]:
# Clustering Neighborhoods
# k-means to cluster the neighborhood into 5 clusters

# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Postcode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 4, 1, 1, 1, 3, 1], dtype=int32)

In [34]:
#create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

# add clustering labels
Toronto_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = df3

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(Toronto_venues_sorted.set_index('Postcode'), on='Postcode')

Toronto_merged.head() # check the last columns!

Postcode          Borough                    Neighborhood   Latitude  \
0      M4E     East Toronto                     The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
3      M4M     East Toronto                 Studio District  43.659526   
4      M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               1          Neighborhood                   Pub   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1           Pizza Place    Italian Restaurant   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               4                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0     Health Food Store           Coffee Shop                   Trail   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2             Pet Store                   Pub           Movie Theater   
3    Italian Restaurant                Bakery     American Restaurant   
4    Dim Sum Restaurant           Swim School             Yoga Studio   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop   Dumpling Restaurant   
1           Yoga Studio             Bookstore               Brewery   
2        Sandwich Place         Burrito Place          Burger Joint   
3           Yoga Studio       Coworking Space    Seafood Restaurant   
4            Donut Shop     Fish & Chips Shop   Filipino Restaurant   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant      Electronics Store  
1              Bubble Tea Shop           Burger Joint  
2         Fast Food Restaurant                Brewery  
3               Sandwich Place            Cheese Shop  
4         Fast Food Restaurant         Farmers Market

In [35]:
map_clusters = folium.Map(location=[lat_tor, lng_tor], zoom_start=11)

# Set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i * x) ** 2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# Add markers to the map
markers_colors = []
for lat, lon, pc, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Postcode'], Toronto_merged['Neighborhood']):
    label = folium.Popup(str(pc) + ', ' + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker([lat, lon],
                        radius=5,
                        popup=('Postcode:' + str('Postcode').capitalize() + '<br>'
                        'Neighborhood: ' + str('Neighborhood')),
                        color='blue',
                        threshold_scale=[0,1,2,3,4],
                        fill=True,
                        fill_color=rainbow,
                        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [36]:
#Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. 
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23  Central Toronto               0                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23      Sushi Restaurant                  Park         Jewelry Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23  Ethiopian Restaurant            Donut Shop   Dumpling Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23  Eastern European Restaurant     Electronics Store            Yoga Studio

In [37]:
# Cluster2 
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               1          Neighborhood   
1       East Toronto               1      Greek Restaurant   
2       East Toronto               1           Pizza Place   
3       East Toronto               1                  Café   
5    Central Toronto               1        Breakfast Spot   
6    Central Toronto               1           Coffee Shop   
7    Central Toronto               1           Pizza Place   
9    Central Toronto               1           Coffee Shop   
11  Downtown Toronto               1                  Park   
12  Downtown Toronto               1           Coffee Shop   
13  Downtown Toronto               1           Coffee Shop   
14  Downtown Toronto               1           Coffee Shop   
15  Downtown Toronto               1                  Café   
16  Downtown Toronto               1           Coffee Shop   
17  Downtown Toronto               1           Coffee Shop   
18  Downtown Toronto               1           Coffee Shop   
19  Downtown Toronto               1           Coffee Shop   
20  Downtown Toronto               1           Coffee Shop   
21  Downtown Toronto               1           Coffee Shop   
24   Central Toronto               1                  Café   
25  Downtown Toronto               1                  Café   
26  Downtown Toronto               1                  Café   
27  Downtown Toronto               1        Airport Lounge   
28  Downtown Toronto               1           Coffee Shop   
29  Downtown Toronto               1           Coffee Shop   
30  Downtown Toronto               1                  Café   
31      West Toronto               1              Pharmacy   
32      West Toronto               1                   Bar   
33      West Toronto               1        Breakfast Spot   
34      West Toronto               1                  Café   
35      West Toronto               1        Breakfast Spot   
36      West Toronto               1           Coffee Shop   
37      East Toronto               1    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue  \
0                             Pub     Health Food Store   
1                     Coffee Shop    Italian Restaurant   
2              Italian Restaurant             Pet Store   
3                     Coffee Shop    Italian Restaurant   
5                             Gym               Dog Run   
6                  Clothing Store   Sporting Goods Shop   
7                  Sandwich Place          Dessert Shop   
9                             Pub           Pizza Place   
11                    Coffee Shop                Bakery   
12            Japanese Restaurant      Sushi Restaurant   
13                           Café                   Pub   
14                 Clothing Store        Cosmetics Shop   
15             Italian Restaurant                 Hotel   
16                   Cocktail Bar            Steakhouse   
17                 Ice Cream Shop    Italian Restaurant   
18                           Café            Steakhouse   
19                          Hotel              Aquarium   
20                           Café                 Hotel   
21                           Café                 Hotel   
24                 Sandwich Place           Coffee Shop   
25            Japanese Restaurant                   Bar   
26  Vegetarian / Vegan Restaurant    Chinese Restaurant   
27                Airport Service      Airport Terminal   
28                           Café            Restaurant   
29                           Café            Steakhouse   
30                  Grocery Store                  Park   
31                    Supermarket                Bakery   
32                    Coffee Shop      Asian Restaurant   
33                    Coffee Shop                  Café   
34             Mexican Restaurant                  Park   
35                      Gift Shop      Cuban Restaurant   
36              

In [38]:
# Cluster 3
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Yoga Studio           Flower Shop           Fish Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22        Farmers Market    Falafel Restaurant            Event Space

In [39]:
# Cluster 4
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8    Central Toronto               3            Playground   
10  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8             Restaurant                   Gym                  Park   
10            Playground              Building                 Trail   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8      Electronics Store      Doner Restaurant            Donut Shop   
10      Doner Restaurant     Fish & Chips Shop   Filipino Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
8    Dumpling Restaurant  Eastern European Restaurant            Event Space  
10  Fast Food Restaurant               Farmers Market     Falafel Restaurant